http/http-advanced-reactive/src/test/java/io/quarkus/ts/http/advanced/reactive/Http2IT.java

```diff
@@ -0,0 +1,259 @@
+        HttpClientOptions httpClientOptions = new HttpClientOptions()
+                .setTrustOptions(new JksOptions().setPath("keystore.jks").setPassword("password"))
+                .setProtocolVersion(HttpVersion.HTTP_2);
+        httpClient = vertx.httpClientBuilder().with(httpClientOptions).build();
+        httpClient.request(HttpMethod.GET, baseUri.getPort(), baseUri.getHost(), BASE_ENDPOINT)
+                .compose(request -> request.send()
+                        .compose(httpClientResponse -> {
+                            assertEquals(HttpStatus.SC_OK, httpClientResponse.statusCode());
+                            assertEquals(HttpVersion.HTTP_2, httpClientResponse.version());
+                            return httpClientResponse.body();
+                        }))
+                .onSuccess(body -> vertxTestContext.verify(() -> {
+                    assertThat("Body response", body.toString().contains(GREETING));
+                    vertxTestContext.completeNow();
+                })).onFailure(Throwable::printStackTrace);
+
+    }
+
+    @Test
+    @DisplayName("Test max header size exceeded error message")
+    @Order(5)
+    void testMaxHeaderSizeExceeded(Vertx vertx, VertxTestContext testContext) {
+        HttpClientOptions options = new HttpClientOptions()
+                .setSsl(true)
+                .setUseAlpn(true)
+                .setTrustOptions(new JksOptions().setPath("keystore.jks").setPassword("password"))
+                .setProtocolVersion(HttpVersion.HTTP_2);
+        httpClient = vertx.httpClientBuilder().with(options).build();
+
+        int expectedLimit = 200;
+        int headerCount;
+
+        for (int i = 0; i < expectedLimit; i++) {
+            headerCount = i;
+
+            try {
+                request(headerCount, testContext);
+            } catch (Exception e) {
+                if (e.getMessage().contains("Header size exceeded max allowed size")) {
+                    testContext.completeNow();
+                    break;
+                } else {
+                    testContext.failNow(e);
+                }
+            }
+        }
+    }
+
+    private static void request(int headerCount, VertxTestContext testContext) {
+
+        httpClient.request(HttpMethod.GET, app.getURI(Protocol.HTTPS).getPort(), baseUri.getHost(), BASE_ENDPOINT)
+                .compose(request -> {
+                    for (int i = 0; i < headerCount; i++) {
+                        String headerName = "Header" + i;
+                        String headerValue = "Value" + i;
+                        request.putHeader(headerName, headerValue);
+                    }
+                    return request.send();
+                })
+                .onFailure(throwable -> {
+                    if (throwable instanceof Http2Exception &&
+                            throwable.getMessage().contains("Header size exceeded max allowed size ")) {
+                        testContext.completeNow();
```

Comment:
Here is how I understand what is happening there:

1. incrementally call Quarkus in asynchronous manner (=when `i` is incremented, the previous request didn't end for sure); first header count is zero
2. when first HTTP request happens, it happens with header count zero, therefore there is no header
3. when first request succeeds, there is no reason to fail, therefore it should succeed, however on success an error is thrown `Unexpected successful response...`.
4. only point at which this test succeed is when header count exceeds maximum limit that you have for `200`

Why does it work? You send 40 HTTP requests that are not rejected by Quarkus because the header limit has not been reached and they are being processed while after that, you send 159 HTTP requests that fail and is rejecting them before processing is faster, you get success result. Am I right?

So my conclusions:

1. You can't reliably say that first 40 requests will be processed after next 159 requests, so there is a potential for flakiness
2. You set `quarkus.http.limits.max-header-list-size=2048` which is reached with `headerCount=41` on my computer. But your test is also successful if you send only `headerCount=90` or any other `headerCount > 40`. Is there a relation between 40 headers and header size limit 2048? My point is that you are testing there is _**some limit applied**_, but you have no idea what that size limit is. If that is the point, IMO you should document it properly with comments.

Please note that number of requests required may differ depending depending on time (future) and environment.

I don't want to force you to rework this here are my thoughts, use it or refuse it:

1. If you don't really care about `2048` limit, you just try to figure if there is a limit, you just need few requests, not sure why there is 200 of them (determine range safe for all envs and future chagnes)
2. You test framework doesn't give a damn about failed results as long as there is at least one successful result (again, I have no idea how the framework works), because I've rewrote your test to this with `headerCount=1`:
```
.onSuccess(response -> {
                    request(150, testContext);
                    System.out.println("response is " + response + " for " + headerCount);
                    testContext.failNow(new AssertionError("Unexpected successful response " + " ; header count is "
                            + headerCount));
                });
```
and test still succeeded even though you clearly can see that `testContext.failNow` is going to be executed (FAIL) and it is most likely executed before `request(150, testContext);` receive response because I/O is more expensive and is executed asynchronously.

I am thinking that this Vert.x framework is probably reliable for one `testContext`, not 200 hundred of them, otherwise how do we know that test is failing?

Reply:
I've simplified the test, thank you Michal for your points.